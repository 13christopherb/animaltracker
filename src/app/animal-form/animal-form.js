import React from 'react';
import {Input, Select} from "../form-fields/form-fields";

/**
 * Form for creating a new animal
 * @param props
 * @returns {*}
 * @constructor
 */
export const AnimalForm = (props) => {
    return (
        <div>
            <div>
                <form onSubmit={props.handleSubmit}>
                    <div className="form-row">
                        <div className="col col-md-2">
                            <Select title={'Location'}
                                    name={'location'}
                                    value={props.location}
                                    onChange={props.handleInputChange}
                                    onBlur={props.onBlur}
                                    options={[
                                        'NRO',
                                        'MBO',
                                        'SLO'
                                    ]}
                            />
                        </div>
                        <div className="col col-md-2">
                            <Select title={'Species'}
                                    name={'species'}
                                    value={props.species}
                                    onChange={props.handleInputChange}
                                    onBlur={props.onBlur}
                                    options={[
                                        'CSL',
                                        'ES',
                                        'HS',
                                        'NFS',
                                        'GFS',
                                        'SSO',
                                        'CET'
                                    ]}
                            />
                        </div>
                        <div className="col col-md-3">
                            <Input id={'name'}
                                   name={'name'}
                                   title={'Name'}
                                   type={'text'}
                                   onChange={props.handleInputChange}
                                   onBlur={props.onBlur}
                                   value={props.name}/>

                        </div>
                        <div className="col col-md-3">
                            <Input id={'weight'}
                                   name={'weight'}
                                   title={'Weight'}
                                   type={'text'}
                                   onChange={props.handleInputChange}
                                   onBlur={props.onBlur}
                                   value={props.weight}/>

                        </div>
                    </div>
                    <div className="form-row">
                        <div className="col col-md-6">
                        <label>
                            Controlled meds:
                            <input
                                name="isGettingControlledMeds"
                                type="checkbox"
                                onChange={props.handleInputChange}
                                onBlur={props.onBlur}
                                value={props.isGettingControlledMeds}
                            />
                        </label>
                        <label>
                            Tube feeding:
                            <input
                                name="isGettingTubed"
                                type="checkbox"
                                onBlur={props.onBlur}
                                onChange={props.handleInputChange}
                                value={props.isGettingTubed}
                            />
                        </label>
                        <button className="btn btn-primary" type="submit">Submit</button>
                        <button onClick={props.toggleAddAnimal} className="btn btn-danger" type="button">
                            Cancel
                        </button>
                        </div>
                    </div>
                </form>
            </div>
        </div>
    )
};

export default AnimalForm