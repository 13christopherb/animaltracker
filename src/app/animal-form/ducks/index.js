import animalFormReducer from './reducers';
export {actions as animalFormActions} from './actions';
export {actionTypes as animalFormTypes} from './action-types';
export default animalFormReducer;