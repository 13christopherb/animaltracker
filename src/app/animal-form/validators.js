/**
 * Contains validators for AnimalForm component
 */

/**
 * Validation function for required values for provided fields
 * @param values Field values to be validated
 */
export const validate = values => {
    const errors = {};
    if (!values.name) {
        errors.name = 'Required';
    } else if (!/^[a-zA-Z\s]+$/.test(values.name)) {
        errors.name = 'Must contain only letters and whitespace'
    } else if (values.name.length > 30) {
        errors.name = 'Must be 30 characters or fewer';
    }
    if (!values.weight) {
        errors.weight = 'Required';
    } else if (!/^\d+(?:\.[0,5]{1})*$/.test(values.weight)) {
        errors.weight = 'Must be an integer or end in .0 or .5';
    }
    if (!values.species) {
        errors.species = 'Required';
    } else if (values.species.length > 3) {
        errors.species = 'Must be fewer than 4 characters'
    }
    if (!values.location) {
        errors.location = 'Required';
    } else if (values.location.length > 3) {
        errors.location = 'Must be fewer than 4 characters'
    }
    return errors
};