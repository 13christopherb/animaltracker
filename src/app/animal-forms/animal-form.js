import React from 'react';
import {Field} from 'redux-form'
import {Form, Col, Button} from 'react-bootstrap'
import {Input, Select} from "../form-fields/index";

/**
 * Form for creating a new animal
 * @param handleSubmit
 * @param handleChange
 * @param onSubmit
 * @param props
 * @returns {*}
 * @constructor
 */
export const AnimalForm = ({handleSubmit, handleChange, onSubmit, ...props}) => {
    return (
        <Form onSubmit={handleSubmit(onSubmit)}>
            <Form.Row>
                <Col md={2}>
                    <Field
                        component={Select}
                        id="location"
                        title="Location"
                        name="location"
                        value={props.value}
                        onChange={handleChange}
                        options={[
                            'NRO',
                            'MBO',
                            'SLO'
                        ]}
                    />
                </Col>
                <Col md={2}>
                    <Field
                        component={Select}
                        id="species"
                        title="Species"
                        name="species"
                        value={props.value}
                        onChange={props.handleChange}
                        options={[
                            'CSL',
                            'ES',
                            'HS',
                            'NFS',
                            'GFS',
                            'SSO',
                            'CET'
                        ]}
                    />
                </Col>
                <Col md={3}>
                    <Field
                        component={Input}
                        id="name"
                        name="name"
                        title="Name"
                        type="text"
                        onChange={handleChange}
                        value={props.value}
                    />
                </Col>
                <Col md={3}>
                    <Field
                        component={Input}
                        id="weight"
                        name="weight"
                        title="Weight"
                        type="text"
                        onChange={handleChange}
                        value={props.value}
                        append="kg"
                    />
                </Col>
            </Form.Row>
            <Form.Row>
                <Col md={8}>
                    <Form.Group>
                        <Form.Label>
                            Controlled meds:
                            <Field
                                component="input"
                                name="isGettingControlledMeds"
                                type="checkbox"
                                onChange={handleChange}
                            />
                        </Form.Label>
                        <Form.Label>
                            Tube feeding:
                            <Field
                                component="input"
                                name="isGettingTubed"
                                type="checkbox"
                                onChange={handleChange}
                            />
                        </Form.Label>
                        <Button variant="primary" disabled={props.pristine || !props.valid} type="submit">
                            Submit
                        </Button>
                        <Button onClick={props.toggleAddAnimal} variant="danger" type="button">
                            Cancel
                        </Button>
                    </Form.Group>
                </Col>
            </Form.Row>
        </Form>
    )
};