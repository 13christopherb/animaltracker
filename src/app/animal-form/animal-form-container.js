import React, {Component} from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import {actions} from './ducks/actions';
import AnimalForm from "./animal-form";


/**
 * Container component for a form to create a new animal
 */
class AnimalFormContainer extends Component {

    constructor(props) {
        super(props);
        this.state = {
            value: '',
            isGettingTubed: false,
            isGettingControlledMeds: false,
            name: '',
            weight: '',
            species: '',
            location: '',
            formErrors: {name: '', weight: '', species: '', location: ''}
        };
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.onBlur = this.onBlur.bind(this)
    }

    componentDidMount() {
        //this.props.dispatch(actions.fetchAnimals());
    }

    handleInputChange(e) {
        const target = e.target;
        const value = target.type === 'checkbox' ? target.checked : target.value;
        const name = target.name;
        this.setState({
            [name]: value
        });
    }

    onBlur(e) {
        const target = e.target;
        const value = target.type === 'checkbox' ? target.checked : target.value;
        const name = target.name;
        this.props.dispatch(actions.changeFieldValue({
            name: name,
            value: value
        }))
    }

    handleSubmit(e) {
        e.preventDefault();
        const animal = {
            timestamp: moment().format(),
            location: this.props.fields.location,
            name: this.props.fields.name,
            species: this.props.fields.species,
            weight: this.props.fields.weight,
            isGettingControlledMeds: this.props.fields.isGettingControlledMeds,
            isGettingTubed: this.props.fields.isGettingTubed
        };
        this.props.dispatch(actions.addAnimal(animal));
    }

    render() {
        return (
            <div>
                <AnimalForm
                    handleInputChange={this.handleInputChange}
                    onBlur={this.onBlur}
                    handleSubmit={this.handleSubmit}
                    location={this.state.location}
                    name={this.state.name}
                    species={this.state.species}
                    weight={this.state.weight}
                    isGettingControlledMeds={this.state.isGettingControlledMeds}
                    isGettingTubed={this.state.isGettingTubed}
                    toggleAddAnimal={this.props.toggleAddAnimal}
                />
            </div>
        );
    }
}

function mapStateToProps({animalForm}, ownProps) {
    return {
        fields: animalForm.fields
    }
}

export default connect(
    mapStateToProps,
)(AnimalFormContainer);