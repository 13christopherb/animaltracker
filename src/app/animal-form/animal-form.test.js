import React from 'react';
import {shallow} from 'enzyme';
import {AnimalForm} from "./animal-form";
import toJson from 'enzyme-to-json';
import moment from 'moment';

const transport = {
    departs: 'MBO',
    arrives: 'NRO',
    meetTime: moment('2017-09-15 09:30:00') //Arbitrary time
};

describe('<AnimalForm />', () => {
    describe('render()', () => {
        test('should render the component', () => {
            const mockSubmit = jest.fn();
            const mockToggle = jest.fn();
            const wrapper = shallow(<AnimalForm
                onSubmit={mockSubmit}
                toggleAddAnimal={mockToggle}
            />);

            expect(toJson(wrapper)).toMatchSnapshot()
        })
    });
});