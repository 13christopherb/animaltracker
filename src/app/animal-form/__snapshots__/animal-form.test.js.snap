// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimalForm /> render() should render the component 1`] = `
<div>
  <div>
    <form
      onSubmit={[MockFunction]}
    >
      <div
        className="form-row"
      >
        <div
          className="col col-md-2"
        >
          <Select
            name="location"
            onChange={[MockFunction]}
            options={
              Array [
                "NRO",
                "MBO",
                "SLO",
              ]
            }
            title="Location"
            value=""
          />
        </div>
        <div
          className="col col-md-2"
        >
          <Select
            name="species"
            onChange={[MockFunction]}
            options={
              Array [
                "CSL",
                "ES",
                "HS",
                "NFS",
                "GFS",
                "SSO",
                "CET",
              ]
            }
            title="Species"
            value=""
          />
        </div>
        <div
          className="col col-md-3"
        >
          <Input
            id="name"
            name="name"
            onChange={[MockFunction]}
            title="Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="col col-md-3"
        >
          <Input
            id="weight"
            name="weight"
            onChange={[MockFunction]}
            title="Weight"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          className="col col-md-6"
        >
          <label>
            Controlled meds:
            <input
              name="isGettingControlledMeds"
              onChange={[MockFunction]}
              type="checkbox"
              value={false}
            />
          </label>
          <label>
            Tube feeding:
            <input
              name="isGettingTubed"
              onChange={[MockFunction]}
              type="checkbox"
              value={false}
            />
          </label>
          <button
            className="btn btn-primary"
            type="submit"
          >
            Submit
          </button>
          <button
            className="btn btn-danger"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
